### PR TITLE
100% coverage

### LINKED ISSUE #29: Get to 100% coverage as reported by coveralls
Coverage decreased post #27. nose does things differently than pure-coverage does.

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #29.

Also simplify the "minimal" environment command by simply skipping tests that can't be run.

### CODE REVIEW

File: docs/testlayer.rst [hunk 1]
```diff
@@ -76,24 +76,29 @@ ZCMLLayer
 
 We now want a layer that loads up some ZCML from a file. The default
 is ``ftesting.zcml``, but here we'll load a test ``testlayer.zcml``.
+We can also choose to provide extra ZCML features (here, "devmode").
```
Comment (tseaver) [inline docs/testlayer.rst]: This probably should either explain what `devmode` does, or provide a link to the docs which do.

File: src/zope/component/standalonetests.py [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
 See: https://bugs.launchpad.net/zope3/+bug/98401
 """
-import sys
-import pickle
 
-def write(x): # pragma: NO COVER
+import sys # pragma: no cover (runs in subprocess)
+import pickle # pragma: no cover (runs in subprocess)
+
+def write(x): # pragma: no cover (runs in subprocess)
```
Comment (tseaver) [inline src/zope/component/standalonetests.py]: ISTM it would be better just to [exclude the whole module via `.configrc`](http://coverage.readthedocs.io/en/coverage-4.2/source.html#execution), e.g.:

```ini
[run]
omit =
    src/zope/component/standalonetests.py
```

File: src/zope/component/testfiles/components.py [hunk 3]
```diff
@@ -54,8 +54,7 @@ def __init__(self, context):
         self.context = context
 
 class Comp3(object):
-    def __init__(self, context):
-        self.context = context
+    pass
```
Comment (tseaver) [inline src/zope/component/testfiles/components.py]: If the constructor is never exercised, then delete the whole class.

File: src/zope/component/tests/examples.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ class ISII(Interface):
     pass
 
 def noop(*args):
-    pass
+    pass # pragma: no cover
```
Comment (tseaver) [inline src/zope/component/tests/examples.py]: Delete it if never called.

File: src/zope/component/tests/examples.py [hunk 5]
```diff
@@ -123,8 +123,7 @@ def __init__(self, context):
 
 @implementer(I3)
 class Comp2(object):
-    def __init__(self, context):
-        self.context = context
+    pass
```
Comment (tseaver) [inline src/zope/component/tests/examples.py]: Delete the class if ctor is never called.

File: src/zope/component/tests/test__api.py [hunk 6]
```diff
@@ -88,8 +88,7 @@ class IBar(Interface):
             pass
         @implementer(IFoo)
         class Global(object):
-            def __init__(self, context):
-                self.context = context
+            pass
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: Maybe have `Global__init__` raise an assertion error instead?  I realize that would require a `pragma:`, but the whole point of the test is that the local registry instantiates `Local` rather than `Global`.

File: src/zope/component/tests/test__api.py [hunk 7]
```diff
@@ -114,10 +113,6 @@ class Test_queryAdapterInContext(unittest.TestCase):
 
     from zope.component.testing import setUp, tearDown
 
-    def _callFUT(self, *args, **kw):
-        from zope.component import queryAdapterInContext
-        return queryAdapterInContext(*args, **kw)
-
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: ISTM that replacing all the imports / direct calls of `queryAdapterInContext` with `self._callFUT` would be smoother.

File: src/zope/component/tests/test__api.py [hunk 8]
```diff
@@ -136,7 +131,7 @@ class Foo(object):
             def __conform__(self, iface, default=None):
                 if iface is IFoo:
                     return _adapted
-                return default
+                raise NotImplementedError()
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: A `pragma:` would be better:  `NotImplementedError` signals a "virtual" method.

File: src/zope/component/tests/test__api.py [hunk 9]
```diff
@@ -148,9 +143,7 @@ class IFoo(Interface):
         _adapted = object()
         class Foo(object):
             def __conform__(self, iface, default=None):
-                if iface is IFoo:
-                    return _adapted
-                return default
+                raise NotImplementedError()
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: Ditto.

File: src/zope/component/tests/test__api.py [hunk 10]
```diff
@@ -338,8 +331,7 @@ class IBar(Interface):
             pass
         @implementer(IFoo)
         class Global(object):
-            def __init__(self, context):
-                self.context = context
+            pass
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: Similar to the case above, an assertion error would be clearer.

File: src/zope/component/tests/test__api.py [hunk 11]
```diff
@@ -570,8 +562,7 @@ class Baz(object):
             pass
         @implementer(IFoo)
         class Global(object):
-            def __init__(self, first, second):
-                self.first, self.second = first, second
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: Here too.

File: src/zope/component/tests/test__api.py [hunk 12]
```diff
@@ -1060,7 +1051,7 @@ def __conform__(self, iface):
             def queryUtility(self, iface, name, default):
                 if iface is IFactory and name == 'test':
                     return _factory
-                return default
+                raise NotImplementedError()
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: `pragma` instead.

File: src/zope/component/tests/test__api.py [hunk 13]
```diff
@@ -1092,7 +1083,7 @@ def __conform__(self, iface):
             def queryUtility(self, iface, name, default):
                 if iface is IFactory and name == 'test':
                     return _Factory()
-                return default
+                raise NotImplementedError()
```
Comment (tseaver) [inline src/zope/component/tests/test__api.py]: `pragma` instead.

File: src/zope/component/tests/test__declaration.py [hunk 14]
```diff
@@ -80,22 +80,11 @@ def _run_generated_code(self, code, globs, locs,
                             fails_under_py3k=True,
                            ):
         import warnings
-        #from zope.component._compat import PYTHON3
-        PYTHON3 = False
```
Comment (tseaver) [inline src/zope/component/tests/test__declaration.py]: This whole testcase class should be wrapped up in a `skipIf(sys.version_info[0] > 2)` check.

File: src/zope/component/tests/test_globalregistry.py [hunk 15]
```diff
@@ -214,7 +219,7 @@ class Foo(object):
             pass
         @adapter(IFoo)
         def _handler(context):
-            assert 0, "DON'T GO HERE"
+            raise NotImplementedError()
```
Comment (tseaver) [inline src/zope/component/tests/test_globalregistry.py]: ISTM that the assert is clearer (with a `pragma:`).

File: src/zope/component/tests/test_globalregistry.py [hunk 16]
```diff
@@ -230,7 +235,7 @@ def test_w_adapts(self):
         class IFoo(Interface):
             pass
         def _handler(context):
-            assert 0, "DON'T GO HERE"
+            raise NotImplementedError()
```
Comment (tseaver) [inline src/zope/component/tests/test_globalregistry.py]: Ditto.

File: src/zope/component/tests/test_hookable.py [hunk 17]
```diff
@@ -68,62 +68,38 @@ def bar():
     def test_cant_assign_original(self):
         from zope.component.hookable import hookable
         def foo():
-            pass
+            raise NotImplementedError()
```
Comment (tseaver) [inline src/zope/component/tests/test_hookable.py]: Wouldn't we be better off just adding `pass` to the list of lines excluded by coverage.

File: docs/testlayer.rst (hunk 1 above)
Comment (jamadden) [inline docs/testlayer.rst]: Are there such docs about "devmode"? 

Maybe just a link to docs about features in general (if there are such things).

File: src/zope/component/tests/test_hookable.py (hunk 17 above)
Comment (jamadden) [inline src/zope/component/tests/test_hookable.py]: In my opinion, no.

I'll try to explain my thinking about the use of `pass` vs `NotImplementedError` in tests.

`pass` is a legitimate thing to write with a specific meaning: This is a function, it *should* be called, although it doesn't actually do anything. If it's not called, coverage will go down, and *something* changed---there's probably a regression.

In contrast, `raise NotImplementedError` has a different but still specific meaning: this is a callable, and I have to give it a name for reasons out of my control, *but it must not* be called. If it is called, *something* changed---there's probably a regression.

I see those two cases as useful distinctions to make. 

I guess the "must not be called" case could usually be spelled "self.fail", although one would have to capture the correct "self" (or the method "self.fail"). Or we could introduce a specific exception class ("MustNotBeCalledError") to raise instead. Or we could have a pre-defined function that we simply assign to the needed name ("foo") in most cases: `foo = callable_that_is_never_called`)...hey, that has the benefit of being a line shorter!

I wish there were a simple way to spell the `pass` (must be called) case that would result in test failures instead of just coverage decreases.

Thoughts?

File: src/zope/component/tests/test_globalregistry.py (hunk 15 above)
Comment (jamadden) [inline src/zope/component/tests/test_globalregistry.py]: Assertions can be turned off. It's not common under test code (I hope) but it is possible.

File: src/zope/component/tests/test__declaration.py (hunk 14 above)
Comment (jamadden) [inline src/zope/component/tests/test__declaration.py]: I don't think that's actually necessary. The `fails_under_py3k` argument can just be removed now...all the tests pass under both versions.

File: src/zope/component/tests/test__api.py (hunk 12 above)
Comment (jamadden) [inline src/zope/component/tests/test__api.py]: Perhaps there should be some assertions (that iface is IFactory and name is test) instead. That way we capture the full intent of the test and break if anything changes.

File: src/zope/component/standalonetests.py (hunk 2 above)
Comment (jamadden) [inline src/zope/component/standalonetests.py]: D'oh. I have no idea why I didn't do that.

File: src/zope/component/tests/test__api.py (hunk 6 above)
Comment (jamadden) [inline src/zope/component/tests/test__api.py]: I'm liking the way the code looks with a `fails_if_called` function:
```python
        @implementer(IFoo)
        class Global(object):
            __init__ = fails_if_called()
```

File: src/zope/component/testfiles/components.py (hunk 3 above)
Comment (jamadden) [inline src/zope/component/testfiles/components.py]: Done. Don't know why it's not showing up as completed.

File: src/zope/component/tests/examples.py (hunk 4 above)
Comment (jamadden) [inline src/zope/component/tests/examples.py]: Done. Don't know why it's not showing as completed.

File: src/zope/component/tests/examples.py (hunk 5 above)
Comment (jamadden) [inline src/zope/component/tests/examples.py]: Done. Don't know why it's not showing as completed.

File: src/zope/component/tests/test__api.py (hunk 11 above)
Comment (jamadden) [inline src/zope/component/tests/test__api.py]: Done.

File: src/zope/component/tests/test_hookable.py (hunk 17 above)
Comment (jamadden) [inline src/zope/component/tests/test_hookable.py]: @tseaver I updated the PR to use the `foo = fails_if_called()` approach instead of `NotImplementedError`. You're right, that does a better job capturing the intent and even reads easier, to me at least.

File: src/zope/component/tests/test__declaration.py (hunk 14 above)
Comment (tseaver) [inline src/zope/component/tests/test__declaration.py]: The "class advice" directives cannot be used at all under Python3.

Comment (jamadden) [inline src/zope/component/tests/test__declaration.py]: That's true for the `zope.interface` versions---they specifically blacklist against that. But `zope.component.adapts` can be used in Python 3:

```pycon
Python 3.5.3 (default, Apr 23 2017, 18:09:27)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from zope.component._declaration import adapts
>>> from zope.interface import Interface
>>> class Foo(object):
...   adapts(Interface)
...
>>> from zope.component._declaration import adaptedBy
>>> adaptedBy(Foo)
(<InterfaceClass zope.interface.Interface>,)
```

Which is the same as under Python 2:
```pycon
Python 2.7.13 (default, May  5 2017, 09:16:47)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from zope.component._declaration import adapts
>>> from zope.interface import Interface
>>> class Foo(object):
...   adapts(Interface)
...
>>> from zope.component._declaration import adaptedBy
>>> adaptedBy(Foo)
(<InterfaceClass zope.interface.Interface>,)
```

So these tests all pass both places.

Comment (jamadden) [inline src/zope/component/tests/test__declaration.py]: For my own reference, because I'd never looked into this before: The `zope.interface` versions are broken because they want to modify the `__metaclass__` variable of the caller, which is what doesn't work on Python 3. They do this in order to arrange to be run *after* the class has been constructed and hence its bases determined, so that they can "flatten" all of the implements declarations from the parent classes into the child class, presumably for performance.

`z.c.adapts` doesn't have this issue, because it just simply assigns its argument to a value in the caller's locals. Parents aren't considered.

File: src/zope/component/tests/test_hookable.py (hunk 17 above)
Comment (mgedmin) [inline src/zope/component/tests/test_hookable.py]: Random drive-by comment: to me `raise AssertionError('should not be called')` seems like the best way to spell must-not-be-called, and `raise NotImplementedError('please override in subclasses')` is the best way to spell abstract-method.

File: docs/testlayer.rst (hunk 1 above)
Comment (mgedmin) [inline docs/testlayer.rst]: There's documentation about features in general: http://zopeconfiguration.readthedocs.io/en/latest/narr.html#making-specific-directives-conditional

`devmode` is a feature that enables additional debugging features for developer convenience, like Django's `DEBUG = True`.  E.g. in my Zope 3 projects I used `hasFeature('devmode')` to decide whether to expose Python tracebacks to the end user in my SystemError views.

Comment (jamadden) [inline docs/testlayer.rst]: That's what I wound up linking to:

> We can also choose to provide extra ZCML features that are used `to
 conditionally control processing of certain directives
 <http://zopeconfiguration.readthedocs.io/en/latest/narr.html#making-specific-directives-conditional>`_
 (here we use "devmode", a common condition for controlling development
 options like debugging output).

File: src/zope/component/tests/test__declaration.py (hunk 14 above)
Comment (tseaver) [inline src/zope/component/tests/test__declaration.py]: OK.  I still think we should be deprecating `adapts` on the `There should be one-- and preferably only one --obvious way to do it.` principle.

**Comment by jamadden** [comment]
Thanks everyone!
